src/Npgsql/TypeHandlers/BitStringHandler.cs

```diff
@@ -71,7 +71,13 @@ public override async ValueTask<BitArray> Read(NpgsqlReadBuffer buf, int len, bo
                     result[bitNo++] = (chunk & (1 << 1)) != 0;
                     result[bitNo++] = (chunk & (1 << 0)) != 0;
                 }
-            } while (bytesLeft > 1);
+
+                if (bytesLeft <= 1)
+                    break;
+
+                if (bytesLeft != 0)
+                    await buf.Ensure(Math.Min(bytesLeft, buf.Size), async);
```

Comment:
buf.Size is the entire size of the buffer: are you sure that's correct? Shouldn't it be the *remaining* size (i.e. `buf.Size - buf.ReadPosition`)?

Reply:
Only the first and last chunks can partly occupy the buffer. Everything between occupies the whole buffer. Right?

Reply:
Let's say bytesLeft is huge, bigger than buf.Size (because the actual bitstring is huge) - that means you're effectively doing `buf.Ensure(buf.Size, async)`. Ensure reads at least X *further* bytes into the buffer, so doing Ensure on buf.Size seems like a mistake... It means you are asking to ensure more bytes than the buffer can hold...

Reply:
Understood and added an assert.

Reply:
There is magin `+1` for the last byte, so your comment was correct.